### ISSUE TITLE
[BUG] page.route with regexp under some conditions does not fullfill

### ISSUE THREAD

**Issue by dein0s** [microsoft/playwright#16918]
**Context:**
- Playwright Version: 1.24.2
- Operating System: Linux
- Node.js version: 16.16.0
- Browser: Chromium

**Code Snippet**

```javascript
const {test, chromium} = require('@playwright/test');

let page;

test.describe('Camera stuff', () => {
  test.beforeEach(async () => {
    const browser = await chromium.launch({
      args: [
        '--disable-gpu',
        '--allow-file-access-from-files',
        '--use-fake-ui-for-media-stream',
        '--use-fake-device-for-media-stream',
        `--use-file-for-fake-video-capture=/some/path/digital-copy.mjpeg`
      ]
    });
    const context = await browser.newContext();
    page = await context.newPage();
    // path to .dat file that is empty; this executes fine multiple times
    await page.route(/\/events\/$/g, route => route.fulfill({path: `/some/path/empty-data.dat`}));
    // path to .json file that contains: {"status": "PENDING"}; this executes only once, if there're multiple requests - i've to copy-paste this route
    await page.route(
      /\/acquisition\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/front\/status\/$/g,
      route => route.fulfill({path: `/some/path/status-PENDING.json`})
    );
  });
  test('do some actions', async () => {
    await page.goto('http://localhost:4200/app/some/url');
    // do some other actions
  });
});

```

**Describe the bug**

Conditions:
- local web app on `localhost:4200`
- local back app on `localhost:8080`

Web app is configured to proceed after multiple `{"status": "PENDING")` responses from back app on `/acquisition/<UUID>/<UUID>/front/status/` endpoint.

While simple regex seems to work as expected and route is always triggered with default settings/options, more complex regex is routed only once (despite `{times: 1}` option is not set). And i have to play a guess game and set those routes (basically copy-paste) multiple times.

Logs with: 1 route, 2 requests, 1 response
```javascript
    await page.route(
      /\/acquisition\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/front\/status\/$/g,
      route => route.fulfill({path: `/some/path/status-PENDING.json`})
    );
 =====
      page.on('request', req => {
        if (req.url().includes('front/status')) {
          console.log(`>> ${req.url()} : this should be routed, check below`);
        }
      });
      page.on('response', async res => {
        try {
          if (res.url().includes('front/status')) {
            console.log(`<< ${res.url()}`);
            console.log(await res.json());
          }
        } catch (err) {}
 =====
>> http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/ : this should be routed, check below
<< http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/
{ status: 'PENDING' }
>> http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/ : this should be routed, check below
...
```
Logs with : 2 routes, 4 requests, 3 responses
```javascript
    await page.route(
      /\/acquisition\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/front\/status\/$/g,
      route => route.fulfill({path: `/some/path/status-PENDING.json`})
    );
    await page.route(
      /\/acquisition\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/front\/status\/$/g,
      route => route.fulfill({path: `/some/path/status-PENDING.json`})
    );
 =====
      page.on('request', req => {
        if (req.url().includes('front/status')) {
          console.log(`>> ${req.url()} : this should be routed, check below`);
        }
      });
      page.on('response', async res => {
        try {
          if (res.url().includes('front/status')) {
            console.log(`<< ${res.url()}`);
            console.log(await res.json());
          }
        } catch (err) {}
 =====
>> http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/ : this should be routed, check below
<< http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/
{ status: 'PENDING' }
>> http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/ : this should be routed, check below
<< http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/
{ status: 'PENDING' }
>> http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/ : this should be routed, check below
<< http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/
{ status: 'PENDING' }
>> http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/ : this should be routed, check below
....
```

Logs with : 3 routes, 4 requests, 4 responses
```javascript
    await page.route(
      /\/acquisition\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/front\/status\/$/g,
      route => route.fulfill({path: `/some/path/status-PENDING.json`})
    );
    await page.route(
      /\/acquisition\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/front\/status\/$/g,
      route => route.fulfill({path: `/some/path/status-PENDING.json`})
    );
    await page.route(
      /\/acquisition\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/[0-9a-fA-F]{8}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{4}\b-[0-9a-fA-F]{12}\/front\/status\/$/g,
      route => route.fulfill({path: `/some/path/status-PENDING.json`})
    );
 =====
      page.on('request', req => {
        if (req.url().includes('front/status')) {
          console.log(`>> ${req.url()} : this should be routed, check below`);
        }
      });
      page.on('response', async res => {
        try {
          if (res.url().includes('front/status')) {
            console.log(`<< ${res.url()}`);
            console.log(await res.json());
          }
        } catch (err) {}
 =====
>> http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/ : this should be routed, check below
<< http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/
{ status: 'PENDING' }
>> http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/ : this should be routed, check below
<< http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/
{ status: 'PENDING' }
>> http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/ : this should be routed, check below
<< http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/
{ status: 'PENDING' }
>> http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/ : this should be routed, check below
<< http://localhost:8080/web/v4/api/acquisition/d005fbe5-6bc8-0000-a000-97b6a7a81aaa/9552ef3b-d414-4ec9-a0f2-2c75b5d081c5/front/status/
{ status: 'PENDING' }

  1 passed (36s)
```

Sadly, i do not have any reproducable repo at this point.

**Comment by pavelfeldman** [comment]
> While simple regex seems to work as expected and route is always triggered with default settings/options, more complex regex is routed only once (despite {times: 1} option is not set)

You are suggesting that the regex matching in Node.js is broken which I find hard to believe. I would encourage you to print more traces using page.on('request') and/or page.route('**/*') to see the entire network traffic and identify the issue on your side. I hope it helps.

**Comment by pavelfeldman** [comment]
I'm still not sure what you are trying to say. Are you saying that not all the requests are processed by page.route? In that case, let's see the list of requests that get into page.route and are reported by page.on('request'). Or are you saying that routes are not passed into all the handlers? That's expected, you have 3 identical routes and the last one fulfills the request, so no other route handlers are called.

**Comment by pavelfeldman** [comment]
We need more information to act on this report. As long as we can't repro it, it is unlikely with can make progress with it. Allow me to close it, but please file a new one and link to this issue when you get back to it!

**Comment by Sommerfeld** [comment]
For people still coming to this thread because they have a related issue:

We also ran into an issue where page.route with a regular expression sometimes didn’t match as expected. This happened even when routes only differed by a single letter.

Oddly, it works if you pass a function that creates a new regex each time, like `(url) => url.match(/.*foo.*/g)`, but not when using a static regex like `/.*foo.*/g`.

In the end this comes down to the global flag and how the route url is matched against the regex.

> GPT: When you use str.match(regexWithGFlag), the lastIndex property of the regex is not used or updated. The match method internally creates its own state, so you can safely reuse the same regex with the g flag in str.match() without odd behavior. However: If you use .exec() or .test() with a regex that has the g (or y) flag, the regex’s lastIndex is updated, which can cause unexpected results if you reuse the same regex object.

**TLDR**: Don't use `g`-flag since you are not interested in multiple matches. But also using `!!str.match(regex)` or `new RegExp(regex).test(str)` instead of `regex.test(str)` (i guess) could make playwright more solid against these mistakes.
